app/javascript/coaches/dashboard/types/coach_dashboard/TimelineEvent.rei

```diff
@@ -1,8 +1,8 @@
 type t;
 
 type reviewResult =
-  | Passed
-  | Failed;
+  | [@dead "reviewResult.Passed"] Passed
+  | [@dead "reviewResult.Failed"] Failed;
```

Comment:
Indeed that's a good point. Even though there are no explicit constructions of values of that type, the construction happens in the implementation of the module and one can get hold of those values indirectly.
There are a couple of things one could do that. One, is to propagate infro from the implementation file, and onbserve that such values are being constructed. Another one is to assume that such values can be constucted, and check if they can flow out of the implementation through the functions defined in the interface.

Reply:
I'm pretty sure that I haven't understood everything that you said correctly because I don't really know how the dead code detection works, but I took another look at the way the interface file and the implementation were written and noticed that it _felt_ wrong.

The type `reviewResult` seemed to be completely independent from the `TimelineEvent.t` _main_ type (module type?). At first, I thought that it could be extracted into a separate module, but all it was being used to do was to differentiate between _passed_ and _failed_ student submissions, so it was just a value computed from the `type t`.

I opted to just get rid the `reviewResult` type and its functions and replaced it with a simple `passed` function that returns a `bool`. The code feels much cleaner now:

```diff
@@ -12,10 +12,6 @@ type t = {
   evaluator: option(string),
 };

-type reviewResult =
-  | Passed
-  | Failed;
-
 let decode = json =>
   Json.Decode.{
     id: json |> field("id", int),
@@ -66,14 +62,8 @@ let reviewPending = tes =>
 let reviewComplete = tes =>
   tes |> List.filter(te => !(te.evaluation |> Grading.pending));

-let getReviewResult = (passGrade, t) =>
-  t.evaluation |> Grading.anyFail(passGrade) ? Failed : Passed;
-
-let resultAsString = reviewResult =>
-  switch (reviewResult) {
-  | Passed => "Passed"
-  | Failed => "Failed"
-  };
+let passed = (~passGrade, t) =>
+  !(t.evaluation |> Grading.anyFail(passGrade));
```

So while the dead-code annotation may not have been technically correct, it did _point_  to a place where _something_ fishy was going on. I just can't clearly define what about the earlier implementation was incorrect - just that the new approach _feels_ cleaner and better.

Reply:
@harigopal it could be a case of the analysis having accidentally spotted a piece of code that could be refactored to look better.
Analyses tend to have very general rules, and one of the difficult things in designing them is resisting the temptation to generalize from a single example.

For example, here's a self-contained example that uses the same kind of pattern:
```
module VariantConstructedOnlyInImplementation: {
  type t =
    | A
    | B; // TODO: discovered this automatically
  let a: t;
} = {
  type t =
    | A
    | B;
  let a = A;
};

let foo = x =>
  switch ((x: VariantConstructedOnlyInImplementation.t)) {
  | A => "this is an A"
  | B => "this is a B"
  };

Js.log(foo(VariantConstructedOnlyInImplementation.a));
```

Reply:
Hmm. The only difference I can see between your example and our previous implementation is that a function in the implementation used both of the constructors...

```reason
let getReviewResult = (passGrade, t) =>
  t.evaluation |> Grading.anyFail(passGrade) ? Failed : Passed;
```

...and this wasn't a _dead_ function. It was used in a component to make the `reviewResult` type, which was then pattern-matched on to render the correct view.

Reply:
If the only pattern matching is in dead functions, then surely exposing the variants is redundant.

The above sentence encodes a more complex rule where dead code and dead type analyses are interconnected, and should be handled carefully. Eg transitivity.

Reply:
Generally there is an increasing level of sophistication in rules one can consider.
As a general point, more sophisticated rules are more powerful but also more difficult to understand. Eg “why is this marked dead and this isn’t, I don’t understand” is the situation one might end up with.
Trade offs...

Reply:
Notice that with non-interactive use the complexity is not important anymore. Though one needs to be really confident that the implementation is correct as you would not want to ship broken code to production.
Interactive use is much more forgiving wrt correctness.

Reply:
I'm sorry, but I think you may have misunderstood me. I still do not understand why the `reviewResult` type was marked dead. This still appears to be a bug in genType's analysis.

You mentioned two rules that `genType` uses to detect dead code:

1. If something is exported to JS from Reason without using genType annotations, it is considered dead code.
2. The analysis is transitive. So if foo is used only by bar and bar is considered dead, then foo is considered dead too.

...plus I'm guessing there's a third basic one:

3. If something is defined, but never referenced elsewhere, then it is considered dead.

If I check with those as the requirements:

1. Inapplicable, because this is not used by JS code.
2. Inapplicable, because there is no _dead_ transitive relationship that can be seen.
3. Inapplicable, because both constructors for the `reviewResult` type are referenced in the implementation file in the `getReviewResult` function, and this function was called in `TimelineEventCard.re` component.

Reply:
The rules I’ve mentioned are not implemented yet.
I’m in the process of revising the implementation. Will get back when done.